File: js/graph.js

class Vertex {
    constructor(name) {
        this.name = name;
        this.conns = [];
    }
}

class Graph {
    constructor() {
        this.vertices = [];
        this.dfsVisited = new Set();
        this.topoSort = new Array();
    }

    addVertex(data) {
        this.vertices.push(new Vertex(data));
    }

    addEdge(src, dest) {
        let srcMatch = this.vertices.filter((vertex) => {
            return vertex.name === src 
         });
         let destMatch = this.vertices.filter((vertex) => {
            return vertex.name === dest 
         });
         srcMatch[0].conns.push(destMatch);
    }
    BFS() {
        const q = [];
        q.push(this.vertices[0]);
        while (q.length > 0) {
            console.log(q[q.length-1]);
            q[q.length-1].conns.forEach((conn) => {
                q.unshift(conn[0]);
            });
            q.pop();
        }
    }
    DFS() {
        this.dfsVisited.add(this.vertices[0].name);
        console.log(this.vertices[0].name);
        this.DFSUtil(this.vertices[0]);
    }
    DFSUtil(curr) {
        if (curr == null) {
            return;
        } else {
            curr.conns.forEach((conn) => {
                if (!this.dfsVisited.has(conn[0].name)) {
                    console.log(conn[0].name);
                    this.dfsVisited.add(conn[0].name);
                    this.DFSUtil(conn[0]);
                }
            });
        }
    }

    CheckCycle() {
        this.dfsVisited.add(this.vertices[0].name);
        console.log(this.vertices[0].name);
        this.CheckCycleUtil(this.vertices[0]);
    }
    CheckCycleUtil(curr) {
        if (curr == null) {
            return;
        } else {
            curr.color = "white";
            curr.conns.forEach((conn) => {
                curr.color = "gray";
                if (conn[0].color == "gray") {
                    console.log(`found a cycle when connecting ${curr.name} and ${conn[0].name}`);
                }
                if (!this.dfsVisited.has(conn[0].name)) {
                    console.log(conn[0].name);
                    this.dfsVisited.add(conn[0].name);
                    this.CheckCycleUtil(conn[0]);
                }
            });
            curr.color = "black";
        }
    }

    TopoSort() {
        this.dfsVisited.add(this.vertices[0].name);
        console.log(this.vertices[0].name);
        this.TopoSortUtil(this.vertices[0]);
        return this.topoSort;
    }
    TopoSortUtil(curr) {
        if (curr == null) {
            return;
        } else {
            curr.color = "white";
            curr.conns.forEach((conn) => {
                curr.color = "gray";
                if (!this.dfsVisited.has(conn[0].name)) {
                    //console.log(conn[0].name);
                    this.dfsVisited.add(conn[0].name);
                    this.TopoSortUtil(conn[0]);
                }
            });
            this.topoSort.push(curr);
            curr.color = "black";
        }
    }
}

let g = new Graph();
g.addVertex("Toronto");
g.addVertex("Markham");
g.addVertex("North Bay");
g.addVertex("Montreal");
g.addVertex("PEI");
g.addVertex("Niagara");
// g.addVertex("Lake"); // Used for cycle
g.addVertex("Costa Rica");

g.addEdge("Toronto", "Markham");
g.addEdge("Toronto", "Montreal");
g.addEdge("Toronto", "Niagara");
g.addEdge("Markham", "North Bay");
// g.addEdge("Montreal", "Lake"); // Used for cycle
g.addEdge("Montreal", "PEI");
// g.addEdge("Lake", "Toronto"); // Used for cycle
g.addEdge("Niagara", "Costa Rica");


// g.DFS();
// g.BFS(); // super basic right now and doesn't include hash of visited

// g.CheckCycle();
const result = g.TopoSort();
console.log('hi');

File: py/graph.py
# next time don't use list for a queue
# has BFS and DFS
class Vertex:
    def __init__(self, data):
        self.data = data
        self.conns = []
    
class Graph:
    def __init__(self):
        self.vertices = []
        self.DFSvisited = []
    
    def addVertex(self, data):
        self.vertices.append(Vertex(data))
    
    def printVertices(self):
        for vertex in self.vertices:
            print(vertex.data)
    
    def addEdge(self, src, dest):
        srcV = [v for v in self.vertices if v.data == src]
        destV = [v for v in self.vertices if v.data == dest]
        srcV[0].conns.append(destV)

    def BFS(self):
        # list for now
        visited = []
        q = []
        q.append(self.vertices[0])
        visited.append(self.vertices[0].data)

        while len(q) > 0:
            current = q.pop(0)
            for child in current.conns:
                if child[0].data not in visited:
                    print(child[0].data)
                    visited.append(child[0].data)
                    q.append(child[0])
    
    def DFS(self):
        self.DFSvisited.append(self.vertices[0].data)
        for conn in self.vertices[0].conns:
            if conn[0].data not in self.DFSvisited:
                self.DFSvisited.append(conn[0].data)
                self._DFS(conn[0])
        self.DFSvisited.clear()

    def _DFS(self, curr):
        print(curr.data)
        for conn in curr.conns:
            self._DFS(conn[0])

g = Graph()

g.addVertex("Toronto")
g.addVertex("Markham")
g.addVertex("Niagara")
g.addVertex("Costa Rica")
g.addVertex("North Bay")
g.addVertex("Montreal")
g.addVertex("PEI")

g.addEdge("Toronto", "Markham")
g.addEdge("Toronto", "Niagara")
g.addEdge("Markham", "North Bay")
g.addEdge("Niagara", "Costa Rica")
g.addEdge("Toronto", "Montreal")
g.addEdge("Montreal", "PEI")

print('==BFS==')
g.BFS()

print('==DFS==')
g.DFS()